page_iterator.go

```diff
@@ -0,0 +1,195 @@
+}
+
+// NewpageIterator creates an iterator instance
+//
+// It has three parameters. res is the graph response from the initial request and represents the first page.
+// reqAdapter is used for getting the next page and constructorFunc is used for serializing next page's response to the specified type.
+func NewPageIterator(res interface{}, reqAdapter GraphRequestAdapterBase, constructorFunc ParsableConstructor) *PageIterator {
+	abstractions.RegisterDefaultSerializer(func() serialization.SerializationWriterFactory {
+		return jsonserialization.NewJsonSerializationWriterFactory()
+	})
+	abstractions.RegisterDefaultDeserializer(func() serialization.ParseNodeFactory {
+		return jsonserialization.NewJsonParseNodeFactory()
+	})
+
+	return &PageIterator{
+		currentPage:     convertToPage(res),
+		reqAdapter:      reqAdapter,
+		pauseIndex:      0,
+		constructorFunc: constructorFunc,
+		headers:         map[string]string{},
+	}
+}
+
+// Iterate traverses all pages and enumerates all items in the current page.
+//
+// Iterate receives a callback function which is called with each item in the current page as an argument. The callback function
+// returns a boolean. To traverse and enumerate all pages always return true and to pause traversal and enumeration
+// return false from the callback.
+//
+// Example
+//      pageIterator := NewPageIterator(resp, reqAdapter, parsableCons)
+//      callbackFunc := func (pageItem interface{}) bool {
+//          fmt.Println(pageitem.GetDisplayName())
+//          return true
+//      }
+//      pageIterator.Iterate(callbackFunc)
+func (pI *PageIterator) Iterate(callback func(pageItem interface{}) bool) {
+	for pI.currentPage != nil {
+		keepIterating := pI.enumerate(callback)
+
+		if !keepIterating {
+			// Callback returned false, stop iterating through pages.
+			return
+		}
+
+		pI.next()
+		pI.pauseIndex = 0 // when moving to the next page reset pauseIndex
+	}
+}
+
+// SetHeaders provides headers for requests made to get subsequent pages
+//
+// Headers in the initial request -- request to get the first page -- are not included in subsequent page requests.
+func (pI *PageIterator) SetHeaders(headers map[string]string) {
+	pI.headers = headers
+}
+
+func (pI *PageIterator) hasNext() bool {
+	if pI.currentPage == nil || pI.currentPage.getNextLink() == nil {
+		return false
+	}
+	return true
+}
+
+func (pI *PageIterator) next() Page {
+	nextPage := pI.getNextPage()
+
+	pI.currentPage = nextPage
+	return nextPage
+}
+
+func (pI *PageIterator) getNextPage() *PageResult {
+	if pI.currentPage.getNextLink() == nil {
+		return nil
+	}
+
+	nextLink, err := url.Parse(*pI.currentPage.getNextLink())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	requestInfo := abstractions.NewRequestInformation()
+	requestInfo.Method = abstractions.GET
+	requestInfo.SetUri(*nextLink)
+	requestInfo.Headers = pI.headers
+
+	res, err := pI.reqAdapter.SendAsync(*requestInfo, pI.constructorFunc, nil)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
could we return the error instead to be consistent with the other APIs offered by the SDK?

Reply:
but the getNextPage function should be able to return the error instead of fatally dying if it runs into an error. And I have mixed opinions about the constructor.

Reply:
Which constructor? **NewPageIterator**?

I wanted execution to stop if I **SendAsync** errors out because that would mean something is seriously wrong. We could return an err, and then set **currentPage** to **nil** so that it fails silently but that results in a hard to debug scenario because we wouldn't know if there's no next page or fetching the next page failed.

Reply:
The decision of whether the error is fatal to the application or not should be left to the person building the application, not us.
In some applications it might be a terrible thing if we can't get the next page (e.g. data processing etc...), and the application should halt. In some other applications it might be an ok thing (e.g. an infinite scroll kind of UI, in which case the application would still run but display an error message).

If we return the error, we're leaving that choice to the person building the application. If we fatally fail, the implementer doesn't have a choice anymore. Does that make sense?